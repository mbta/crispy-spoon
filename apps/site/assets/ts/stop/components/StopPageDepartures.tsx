import { filter, groupBy, sortBy } from "lodash";
import React, { ReactElement, useState } from "react";
import { Alert, Route } from "../../__v3api";
import DeparturesFilters, { ModeChoice } from "./DeparturesFilters";
import { isRailReplacementBus, modeForRoute } from "../../models/route";
import DepartureCard from "./DepartureCard";
import { alertsByRoute, isInNextXDays } from "../../models/alert";
import { DepartureInfo } from "../../models/departureInfo";
import { GroupedRoutePatterns } from "../../models/route-patterns";

interface StopPageDeparturesProps {
  routes: Route[];
  departureInfos: DepartureInfo[];
  alerts: Alert[];
  groupedRoutePatterns: GroupedRoutePatterns;
}

// Commuter Rail, then Subway, then Bus
const modeSortFn = ({ type }: Route): number => {
  if (type === 2) {
    return 1;
  }
  if (type in [0, 1]) {
    return 2;
  }
  return type;
};

const StopPageDepartures = ({
  routes,
  departureInfos,
  alerts,
  groupedRoutePatterns
}: StopPageDeparturesProps): ReactElement<HTMLElement> => {
  // default to show all modes.
  const [selectedMode, setSelectedMode] = useState<ModeChoice>("all");
  // Filters our all replacement buses that haven't been remapped in the parent component
  const nonShuttleRoutes = filter(routes, r => !isRailReplacementBus(r));
  const groupedRoutes = groupBy(nonShuttleRoutes, modeForRoute);
  const modesList = Object.keys(groupedRoutes) as ModeChoice[];
  const filteredRoutes =
    selectedMode === "all" ? nonShuttleRoutes : groupedRoutes[selectedMode];
  const currentAlerts = filter(alerts, a => isInNextXDays(a, 0));
  const groupedAlerts = alertsByRoute(currentAlerts);

  return (
    <div className="routes">
      {modesList.length > 1 && (
        <DeparturesFilters
          modesList={modesList}
          selectedMode={selectedMode}
          setSelectedMode={setSelectedMode}
        />
      )}
      <ul className="stop-departures list-unstyled">
        {sortBy(filteredRoutes, [modeSortFn, "sort_order"]).map(route => (
          <DepartureCard
            key={route.id}
            route={route}
            routePatternsByHeadsign={groupedRoutePatterns[route.id]}
            alertsForRoute={groupedAlerts[route.id] || []}
            departuresForRoute={departureInfos.filter(
              d => d.route.id === route.id
            )}
          />
        ))}
      </ul>
    </div>
  );
};

export default StopPageDepartures;
