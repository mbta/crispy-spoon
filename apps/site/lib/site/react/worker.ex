defmodule Site.React.Worker do
  @moduledoc """
  React renderer worker
  """
  require Logger
  use GenServer

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts)
  end

  def init(_opts) do
    node = System.find_executable("node") || System.find_executable("nodejs")
    config = Application.get_env(:site, :react)
    port = Port.open({:spawn_executable, node}, args: [config[:build_path]])
    {:ok, %{port: port}}
  end

  def handle_call({:render, name, props}, _from, %{port: port} = state) do
    body =
      Poison.encode!(%{
        name: name,
        props: props
      })

    Port.command(port, body <> "\n")

    response =
      ""
      |> receive_response(&handle_json/1)

    case response do
      %{data: %{"error" => nil} = data} ->
        {:reply, {:ok, data}, state}

      %{data: %{"error" => _error} = data} ->
        {:reply, {:error, data}, state}

      _ ->
        {:noreply, state}
    end
  end

  defp receive_response(prev_responses, handle_fn) do
    receive do
      {_, {:data, data}} ->
        str = to_string(data)
        resp = prev_responses <> str
        handle_response(resp, handle_fn)
    end
  end

  def handle_response(resp, handle_fn) do
    # Split and keep delimiter
    ~r/(?<=\n)/
    |> Regex.split(resp, trim: true)
    |> Enum.map(&handle_data(&1, handle_fn))
    |> Enum.find(&Map.has_key?(&1, :data))
  end

  def handle_data(str, handle_fn) do
    if String.ends_with?(str, "\n") do
      if String.starts_with?(str, "node_logging") do
        handle_logging(str)
      else
        handle_fn.(str)
      end
    else
      receive_response(str, handle_fn)
    end
  end

  def handle_json(msg) do
    json = Poison.decode!(msg)
    %{data: json}
  end

  def handle_logging(msg) do
    _ = Logger.warn(inspect(msg))
    %{}
  end

  def handle_info(msg, state) do
    _ = Logger.warn(inspect(msg))
    {:noreply, state}
  end
end
