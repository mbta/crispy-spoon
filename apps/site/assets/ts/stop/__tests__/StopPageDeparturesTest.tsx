import React from "react";
import { render, screen } from "@testing-library/react";
import StopPageDepartures from "../components/StopPageDepartures";
import { Route, RouteType } from "../../__v3api";

const baseRoute = (name: string, type: RouteType): Route =>
  ({
    id: name,
    direction_destinations: { 0: "Somewhere there", 1: "Over yonder" },
    name: `${name} Route`,
    type
  } as Route);
const routeData: Route[] = [baseRoute("4B", 3), baseRoute("Magenta", 1)];
const mockClickAction = jest.fn();

describe("StopPageDepartures", () => {
  it("renders with no data", () => {
    const { asFragment } = render(
      <StopPageDepartures
        routes={[]}
        onClick={mockClickAction}
        stopName=""
        alerts={[]}
        departureInfos={[]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole("list")).toBeEmptyDOMElement();
  });

  it("renders with data", () => {
    const { asFragment } = render(
      <StopPageDepartures
        routes={routeData}
        onClick={mockClickAction}
        stopName=""
        alerts={[]}
        departureInfos={[]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByRole("list")[0]).not.toBeEmptyDOMElement();
    ["All", "Bus", "Subway"].forEach(name => {
      expect(
        screen.getByRole("button", { name: new RegExp(name) })
      ).toBeDefined();
    });
  });

  it("doesn't show the filters if there is 1 mode present", () => {
    render(
      <StopPageDepartures
        routes={[routeData[0]]}
        onClick={() => {}}
        stopName=""
        alerts={[]}
        departureInfos={[]}
      />
    );
    expect(
      screen.queryByRole("button", { name: new RegExp("All") })
    ).toBeNull();
  });
});
