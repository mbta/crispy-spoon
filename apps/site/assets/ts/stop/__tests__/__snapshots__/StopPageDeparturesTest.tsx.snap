// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopPageDepartures renders with data 1`] = `<DocumentFragment />`;

exports[`StopPageDepartures renders with data 2`] = `
<DocumentFragment>
  <div
    class="routes"
  >
    <div
      class="departure-filters"
    >
      <button
        class="
            btn btn-secondary active
        "
        type="button"
      >
         All
      </button>
      <button
        class="
            btn btn-secondary inactive
        "
        type="button"
      >
        <span
          aria-hidden="false"
          class="notranslate c-svg__icon"
        >
          SVG
        </span>
         Subway
      </button>
      <button
        class="
            btn btn-secondary inactive
        "
        type="button"
      >
        <span
          aria-hidden="false"
          class="notranslate c-svg__icon"
        >
          SVG
        </span>
         Bus
      </button>
    </div>
    <ul
      class="stop-departures list-unstyled"
    >
      <li
        class="departure-card"
      >
        <a
          class="departure-card__route u-bg--red-line notranslate"
          data-turbolinks="false"
          href="/schedules/Red"
        >
          <span
            aria-hidden="true"
            class="notranslate c-svg__icon"
          >
            SVG
          </span>
           
          <span
            class=""
          >
            Red Route
          </span>
        </a>
        <div
          aria-label="Open upcoming departures to Braintree"
          class="departure-card__headsign d-flex"
          role="button"
          tabindex="0"
        >
          <div
            class="departure-card__headsign-name notranslate"
          >
            Braintree
          </div>
          <div
            class="departure-card__content"
          >
            <div
              class="font-helvetica-neue fs-14 u-nowrap"
            >
              No more trips today
            </div>
          </div>
          <div>
            <i
              aria-hidden="true"
              class="fa fa-angle-right notranslate fa-fw"
            />
          </div>
        </div>
        <div
          aria-label="Open upcoming departures to Ashmont"
          class="departure-card__headsign d-flex"
          role="button"
          tabindex="0"
        >
          <div
            class="departure-card__headsign-name notranslate"
          >
            Ashmont
          </div>
          <div
            class="departure-card__content"
          >
            <div
              class="font-helvetica-neue fs-14 u-nowrap"
            >
              No more trips today
            </div>
          </div>
          <div>
            <i
              aria-hidden="true"
              class="fa fa-angle-right notranslate fa-fw"
            />
          </div>
        </div>
        <div
          aria-label="Open upcoming departures to Alewife"
          class="departure-card__headsign d-flex"
          role="button"
          tabindex="0"
        >
          <div
            class="departure-card__headsign-name notranslate"
          >
            Alewife
          </div>
          <div
            class="departure-card__content"
          >
            <div
              class="font-helvetica-neue fs-14 u-nowrap"
            >
              No more trips today
            </div>
          </div>
          <div>
            <i
              aria-hidden="true"
              class="fa fa-angle-right notranslate fa-fw"
            />
          </div>
        </div>
      </li>
      <li
        class="departure-card"
      >
        <a
          class="departure-card__route u-bg--bus notranslate"
          data-turbolinks="false"
          href="/schedules/16"
        >
          <span
            aria-hidden="true"
            class="notranslate c-svg__icon"
          >
            SVG
          </span>
           
          <span
            class="bus-route-sign"
          >
            16 Route
          </span>
        </a>
        <div
          aria-label="Open upcoming departures to Harbor Point"
          class="departure-card__headsign d-flex"
          role="button"
          tabindex="0"
        >
          <div
            class="departure-card__headsign-name notranslate"
          >
            Harbor Point
          </div>
          <div
            class="departure-card__content"
          >
            <div
              class="font-helvetica-neue fs-14 u-nowrap"
            >
              No more trips today
            </div>
          </div>
          <div>
            <i
              aria-hidden="true"
              class="fa fa-angle-right notranslate fa-fw"
            />
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`StopPageDepartures renders with no data 1`] = `<DocumentFragment />`;

exports[`StopPageDepartures renders with no data 2`] = `
<DocumentFragment>
  <div
    class="routes"
  >
    <ul
      class="stop-departures list-unstyled"
    />
  </div>
</DocumentFragment>
`;
